alf/algorithms/algorithm.py

```diff
@@ -156,28 +159,71 @@ def get_optimizer_and_module_sets(self):
         def _is_alg(obj):
             return isinstance(obj, Algorithm)
 
-        def _is_module_or_var(obj):
-            return ((isinstance(obj, tf.Variable) and obj.trainable)
-                    or (isinstance(obj, tf.Module)
-                        and not isinstance(obj, Algorithm)))
+        def _is_trainable_module(obj):
+            """Only return True if the module or var is trainable, to avoid
+            possible confusions shown in the optimizer info"""
+            return (isinstance(obj, tf.Module)
+                    and not isinstance(obj, Algorithm)
+                    and obj.trainable_variables)
+
+        def _is_trainable_var(obj):
+            """Only return True if the module or var is trainable, to avoid
+            possible confusions shown in the optimizer info"""
+            return isinstance(obj, tf.Variable) and obj.trainable
+
+        def _get_trainable_vars(obj):
+            if obj is None:
+                return []
+            if _is_trainable_module(obj):
+                return obj.trainable_variables
+            elif _is_trainable_var(obj):
+                return [obj]
+            else:
+                assert "%s is not a trainble module or variable!" % obj.name
 
         module_sets = [copy.copy(s) for s in self._init_module_sets]
         optimizers = copy.copy(self._init_optimizers)
-        module_ids = set(map(id, sum(module_sets, [])))
+        algorithm_names = [self.__class__.__name__] * len(optimizers)
+
+        # This set stores all the seen distinct variables so far in this alg
+        var_ids = set(
+            map(
+                id,
+                tf.nest.flatten(
+                    tf.nest.map_structure(_get_trainable_vars, module_sets))))
 
         for alg in self._get_children(_is_alg):
-            for opt, module_set in alg.get_optimizer_and_module_sets():
+            opt_and_module, child_var_ids = alg.get_optimizer_and_module_sets()
```

Comment:
This function name as well as its documentation might need to be updated: https://github.com/HorizonRobotics/alf/blob/d58103f299558f5fdeae218bd23515aed5c9b557/alf/algorithms/algorithm.py#L147-L153

Reply:
The name is fine since var_ids is just an auxiliary variable to assist getting optimizers and modules.